### PR TITLE
implements management of creation cluster status conditions

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/pull/2209.

### CODE REVIEW

File: create.go [hunk 1]
```diff
@@ -190,6 +221,22 @@ func (r *Resource) computeCreateEventPatches(ctx context.Context, accessor metav
...
+		sameVersion := allNodesHaveVersion(clusterStatus.Nodes, desiredVersion)
+
+		if isUpdated && sameCount && sameVersion {
+			patches = append(patches, Patch{
+				Op:    "replace",
+				Path:  "/status/cluster/versions",
+				Value: clusterStatus.WithNewVersion(desiredVersion),
+			})
+		}
+	}
```
Comment (xh3b4sd) [inline create.go]: This got just moved more to the bottom so the condition management above is a bit more telling about the natural status transitions.

File: create.go [hunk 2]
```diff
@@ -255,7 +302,6 @@ func (r *Resource) computeCreateEventPatches(ctx context.Context, accessor metav
 	}
 
 	// TODO emit metrics when update did not complete within a certain timeframe
-	// TODO update status condition when guest cluster is migrating from creating to created status
```
Comment (xh3b4sd) [inline create.go]: This is done with this PR.

File: create.go [hunk 3]
```diff
@@ -177,11 +208,11 @@ func (r *Resource) computeCreateEventPatches(ctx context.Context, accessor metav
...
 	// indicating the guest cluster is updating now.
 	{
-		isNotEmpty := currentVersion != ""
-		isNotUpdating := !clusterStatus.HasUpdatingCondition()
+		isCreated := clusterStatus.HasCreatedCondition()
+		notUpdating := !clusterStatus.HasUpdatingCondition()
 		versionDiffers := currentVersion != desiredVersion
 
-		if isNotEmpty && isNotUpdating && versionDiffers {
+		if isCreated && notUpdating && versionDiffers {
```
Comment (xh3b4sd) [inline create.go]: The version and updating status management was combined and is now separated. Should be more reliable and independent from each other.

File: create.go [hunk 4]
```diff
@@ -87,7 +87,7 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	return nil
 }
 
-func (r *Resource) computeCreateEventPatches(ctx context.Context, accessor metav1.Object, obj interface{}) ([]Patch, error) {
+func (r *Resource) computeCreateEventPatches(ctx context.Context, obj interface{}) ([]Patch, error) {
```
Comment (xh3b4sd) [inline create.go]: I noticed the accessor is not used at all for the patch computation.

**Review by MarcelMue — APPROVED** [review]
This is LGTM.

**Review by rossf7 — APPROVED** [review]
LGTM

File: create.go [hunk 5]
```diff
@@ -150,25 +150,56 @@ func (r *Resource) computeCreateEventPatches(ctx context.Context, accessor metav
...
+				Value: clusterStatus.WithCreatingCondition(),
+			})
+		}
+	}
+
+	// Once the guest cluster is created we set the according status condition so
+	// the cluster status reflects the transitioning from creating to created.
+	{
+		isCreating := clusterStatus.HasCreatingCondition()
+		notCreated := !clusterStatus.HasCreatedCondition()
```
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply create.go]: This should not happen. I have an issue during testing though right now where some conditions cause other conditions to kick in and we get an endless loop. Have to check on this.

**Pull request merged**